src/common/polar_decomposition.h

```diff
@@ -0,0 +1,380 @@
+  BB[1][2] = BB[2][1];
+
+  BB[3][1] = BB[3][1] - L[1][0] * BB[0][3];
+  BB[1][3] = BB[3][1];
+  BB[2][2] = BB[2][2] - L[2][0] * BB[0][2];
+
+  BB[3][2] = BB[3][2] - L[2][0] * BB[0][3];
+  BB[2][3] = BB[3][2];
+  BB[3][3] = BB[3][3] - L[3][0] * BB[0][3];
+
+  // Second step
+  r = 2;
+  if(BB[2][2] < BB[1][1]) r = 1;
+  if(BB[r][r] > BB[0][0])
+  {
+    // p([2 r(1)]) = p([r(1) 2]);
+    float p_r = p[r];
+    float p_1 = p[1];
+    p[r] = p_1;
+    p[1] = p_r;
+
+    // BB([2 r],:) = BB([r 2],:);
+    float BB_1[4] = { BB[1][0], BB[1][1], BB[1][2], BB[1][3] };
+    float BB_r[4] = { BB[r][0], BB[r][1], BB[r][2], BB[r][3] };
+    for(size_t j = 0; j < 4; j++)
+    {
+      // invert lines 1 and r
+      BB[1][j] = BB_r[j];
+      BB[r][j] = BB_B[j];
+    }
+
+
+    BB(:,[1 r]) = BB(:,[r 1]);
+    L([1 r(0)],:) = L([r(0) 1],:);
+    L(:,[1 r(0)]) = L(:,[r(0) 1]);
+  }
+
+
+  D(1,1) = BB(1,1);
+  L(2,1) = BB(2,1)/D(1,1);
+  L(3,1) = BB(3,1)/D(1,1);
+  D(2,2) = BB(2,2)-L(2,1)*BB(1,2);
+  D(3,2) = BB(3,2)-L(2,1)*BB(1,3);
+  D(2,3) = D(3,2);
+  D(3,3) = BB(3,3)-L(3,1)*BB(1,3);
+
+  DD = D(3,3)*D(4,4)-D(3,4)*D(3,4);
+  if DD == 0, %treat specially
+      if max(abs(D(3:4,3:4))) == 0, v = [L(2,1)*L(4,2)-L(4,1);-L(4,2);0;1];v = v/norm(v);
+      else
+          v = L'\[0;0;null(D(3:4,3:4))];v = v/norm(v);
+      end
+  else
+  ID = [D(4,4) -D(3,4); -D(3,4) D(3,3)];
+
+    if subspa
+        v = [L(2,1)*L(3,2)-L(3,1) L(2,1)*L(4,2)-L(4,1);-L(3,2) -L(4,2);1 0;0 1];
+        IL = [1 0 0 0;-L(2,1) 1 0 0;v'];
+        [v ~] = qr(v,0);%->cost in flops if implemented by hand: 37 M+24 A+4 O
+        v = IL*v;%it looks faster to multiply than to solve lin syst (even though should be same flops)
+        v(1,:) = v(1,:)/D(1,1);
+        v(2,:) = v(2,:)/D(2,2);
+        v(3:4,:) = ID*v(3:4,:)/DD(1);
+        v = v'*IL;v = v';
+        v = IL*v;
+        v(1,:) = v(1,:)/D(1,1);
+        v(2,:) = v(2,:)/D(2,2);
+        v(3:4,:) = ID*v(3:4,:)/DD(1);
+        v = v'*IL;v = v';
+                [v ~] = qr(v,0);
+        H = v'*L;H = -H*D*H';%Cheaper
+        if abs(H(1,2))<1e-15
+            if H(1,1)>H(1,2)
+                v = v(:,1);
+            else
+                v = v(:,2);
+            end
+        else
+            r = (H(1,1)-H(2,2))/(2*H(1,2));
+            v = v*[r+sign(H(1,2))*sqrt(1+r(1)*r(1));1];
+            v = v/norm(v);
+        end
+
+    else
+        v = [L(2,1)*L(4,2)+L(3,1)*L(4,3)-L(2,1)*L(4,3)*L(3,2)-L(4,1);
+             L(4,3)*L(3,2)-L(4,2); -L(4,3) ;1];
+        IL = [1 0 0 0; -L(2,1) 1 0 0; L(2,1)*L(3,2)-L(3,1) -L(3,2) 1 0; v'];
+        nv = realsqrt(v(1)*v(1)+v(2)*v(2)+v(3)*v(3)+v(4)*v(4));
+        v = v/nv(1);
+
+        for it = 1:nit
+            v = IL*v;
+            v(1) = v(1)/D(1,1);v(2) = v(2)/D(2,2);v(3:4) = ID*v(3:4)/DD(1);
+            v = v'*IL;v = v';
+            nv = realsqrt(v(1)*v(1)+v(2)*v(2)+v(3)*v(3)+v(4)*v(4));
+            v = v/nv(1);
+        end
+    end
+    end
+    v(p) = v;
```

Comment:
Everything goes nice, suddenly... Pythons! Bunch of them! :stuck_out_tongue_closed_eyes:

Reply:
WIP. It's actually Matlab code.

Reply:
Ah shit... i haven't noticed lack of ':' and misaligned `end` ;)